### PR TITLE
docs: auth passwords page

### PR DESCRIPTION
**Pull request by hf**
Refactors the whole Password Reset page into a Passwords page that also explains the password reset flow.

### CODE REVIEW

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+Passwords derive their security from how easy they are to guess or brute-force In theory this means that the longer a password is and the more required characters it has (digits, lowercase and uppercase letters, symbols) the harder it will be to guess.
+
+This table shows the _minimum_ number of guesses that need to be tried to access a user's account:
+
+| Required characters                          | Length | Guesses          |
+| -------------------------------------------- | ------ | ---------------- |
+| Digits only                                  | 8      | ~ 2<sup>27</sup> |
+| Digits and letters                           | 8      | ~ 2<sup>39</sup> |
+| Digits, lower and uppercase letters          | 8      | ~ 2<sup>48</sup> |
+| Digits, lower and uppercase letters, symbols | 8      | ~ 2<sup>52</sup> |
```
Comment (hf) [inline apps/docs/pages/guides/auth/passwords.mdx:23]: Basically `log2((characters)^(length))`...

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 2]
```diff
@@ -0,0 +1,155 @@
+import Layout from '~/layouts/DefaultGuideLayout'
+import { Accordion } from 'ui'
+
+export const meta = {
+  id: 'auth-passwords',
+  title: 'Passwords',
+  description: 'How to work with passwords in Supabase Auth',
+}
+
+Using passwords to authenticate users is a tried-and-tested method to give your users access to your application. Supabase Auth provides you with secure configuration options and uses best-practices to store and verify your user's passwords.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:10]:
```suggestion
Using passwords to authenticate users is a tried-and-tested method to give your users access to your application. Supabase Auth provides you with secure configuration options and uses best practices to store and verify your user's passwords.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+## Password security
+
+Passwords derive their security from how easy they are to guess or brute-force In theory this means that the longer a password is and the more required characters it has (digits, lowercase and uppercase letters, symbols) the harder it will be to guess.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:14]:
```suggestion
A password is more secure if it is harder to guess or brute-force. In theory, a password is harder to guess if it is longer. It is also harder to guess if it uses a larger set of characters (for example, digits, lowercase and uppercase letters, and symbols).
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+This table shows the _minimum_ number of guesses that need to be tried to access a user's account:
+
+| Required characters                          | Length | Guesses          |
+| -------------------------------------------- | ------ | ---------------- |
+| Digits only                                  | 8      | ~ 2<sup>27</sup> |
+| Digits and letters                           | 8      | ~ 2<sup>41</sup> |
+| Digits, lower and uppercase letters          | 8      | ~ 2<sup>48</sup> |
+| Digits, lower and uppercase letters, symbols | 8      | ~ 2<sup>52</sup> |
+
+In reality though, passwords are not always generated at random, but often contain variations of names, words, dates, sentences or phrases. Malicious actors can use these properties to guess a password in less attempts.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:25]:
```suggestion
In reality though, passwords are not always generated at random. They often contain variations of names, words, dates, and common phrases. Malicious actors can use these properties to guess a password in fewer attempts.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+There are hundreds of millions (and growing!) known passwords out there. Malicious actors can use these lists of leaked passwords to automate login attempts (known as credential stuffing) and steal or access sensitive user data.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:27]: Also, passwords are often reused. Malicious actors can use lists of leaked passwords to automate login attempts (known as credential stuffing).

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+### Password strength and leaked password protection
+
+To help protect your users data and accounts, Supabase Auth allows you fine-grained control over the strength of the passwords used on your project. You can configure these in your project's [Auth settings](/dashboard/project/_/settings/auth):
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:31]:
```suggestion
To help protect your users, Supabase Auth allows you fine-grained control over the strength of the passwords used on your project. You can configure these in your project's [Auth settings](/dashboard/project/_/settings/auth):
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+- Set a large minimum password length. Anything less than 8 characters is not recommended.
+- Set the required characters that must appear at least once in a user's password to be accepted. Use the strongest option of requiring digits, lowercase and uppercase letters and symbols.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:34]:
```suggestion
- Set the required characters that must appear at least once in a user's password. Use the strongest option of requiring digits, lowercase and uppercase letters, and symbols.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+- Prevent the use of leaked passwords. Supabase Auth uses the open-source [HaveIBeenPwned.org Pwned Passwords API](https://haveibeenpwned.com/Passwords) to reject passwords that have been leaked and are known by malicious actors.
+
+### Additional recommendations
+
+In addition to choosing suitable password strength settings and preventing the use of leaked passwords, consider prominently asking your users to:
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:39]:
```suggestion
In addition to choosing suitable password strength settings and preventing the use of leaked passwords, consider asking your users to:
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+- Use a password manager to store and generate passwords.
+- Avoid password reuse across websites and apps.
+- Avoid using personal information in passwords.
+- Use [Multi-Factor Authentication](/docs/guides/auth/auth-mfa) whenever possible.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:44]:
```suggestion
- Use [Multi-Factor Authentication](/docs/guides/auth/auth-mfa).
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+## Resetting a user's password (forgot password)
+
+Strong passwords are difficult to remember, so Supabase Auth provides you with APIs to build a secure password reset flow.
+
+### Overview
+
+To add a secure password reset flow in your application, follow these steps.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:52]:
```suggestion
To add a secure password reset flow to your application:
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+1. Build the **password reset page**:
+    - Should be publicly accessible and contain a form asking for the user's email address.
+    - Use the [`supabase.auth.resetPasswordForEmail`](/docs/reference/javascript/auth-resetpasswordforemail) API to request a password reset link for a user's email address.
+    - Specify the `redirectTo` parameter when calling this API to point to the URL of the **change password page.**
+2. Build the **password change page**:
+    - Should be accessible only to authenticated users.
+    - Add its URL to the allowed [Redirect URLs](/dashboard/project/_/auth/url-configuration) settings.
+    - Show a form or other propmt asking the user to choose a new password.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:61]:
```suggestion
    - Show a form or other prompt asking the user to choose a new password.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    - Call the [`supabase.auth.updateUser`](/docs/reference/javascript/auth-updateuser) API to set a new password for the user.
+
+The email link sent to your users works very similarly to [passwordless authentication](/docs/guides/auth/passwordless-login):
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:64]:
```suggestion
The email link sent to your users works in the same way as [passwordless authentication](/docs/guides/auth/passwordless-login):
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+1. After a user visits the **reset password page** they are sent a reset password link.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:66]:
```suggestion
1. After a user visits the **reset password page**, they are sent a reset password link.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    - If you use PKCE (default) this link will be bound to the browser or device the user asked for password reset. Consider informing the user in the email message to use the same device or browser they requested the link from.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:67]:
```suggestion
    - If you use PKCE (default), this link only works on the device or browser where the original reset request was made. Display a message to the user to make sure they don't change devices or browsers.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+    - If you use the implicit grant flow, the link will be valid when opened on any device.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:68]:
```suggestion
    - If you use the implicit grant flow, the link can be opened on any device.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+2. A user clicks the link and is taken first to Supabase Auth which validates it.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:69]:
```suggestion
2. A user clicks the link. They are taken to Supabase Auth, which validates the link.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+    - If the link is valid, it will redirect to the **change password page** which was specified in the `redirectTo` parameter.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:70]:
```suggestion
    - If the link is valid, it redirects to the **change password page**, which you specified in the `redirectTo` parameter.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+    - If the **change password page's** URL is not properly registered in the [Redirect URLs](/dashboard/project/_/auth/url-configuration) configuration, the user will be taken to the default Site URL page.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:71]:
```suggestion
    - If the **change password page's** URL is not properly registered in the [Redirect URLs](/dashboard/project/_/auth/url-configuration) configuration, the user is taken to the default Site URL page.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+3. Supabase Auth redirects to the **change password page** including [session](/docs/guides/auth/sessions) information in the URL.
+    - If you used PKCE (default) when the user asked for a reset password link, the redirect will contain the `code` query param.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:73]:
```suggestion
    - If you used PKCE (default), the redirect contains the `code` query param.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+    - If you are not using official Supabase libraries, or have a custom setup, extract the `code` from the URL and call the [`supabase.auth.exchangeCodeForSession`](/docs/reference/javascript/auth-exchangecodeforsession) API.
+    - In most cases using official Supabase libraries will handle this for you.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:75]:
```suggestion
    - Official Supabase libraries handle code the code exchange for you.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+    - If you used the implicit flow when the user asked for a reset password link, the redirect will contain a URL fragment encoding the user's [session](/docs/guides/auth/sessions).
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:76]:
```suggestion
    - If you used the implicit flow, the redirect contains a URL fragment encoding the user's [session](/docs/guides/auth/sessions).
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+<Tabs
+  scrollable
+  size="small"
+  type="underlined"
+  defaultActiveId="js"
+  queryGroup="language"
+>
+<TabPanel id="js" label="JavaScript">
+
+Supabase provides a convenient method [`supabase.auth.resetPasswordForEmail`](/docs/reference/javascript/auth-resetpasswordforemail) to reset a user password. This method takes a parameter of `redirectTo` which we will use to pass an absolute URL to the update password page. This URL must be saved in your allowed [Redirect URLs](https://supabase.com/dashboard/project/_/auth/url-configuration) list found at [Authentication > Redirect Configuration](https://supabase.com/dashboard/project/_/auth/url-configuration) or it won't redirect the user.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:89]:
```suggestion
Supabase provides a convenient method [`supabase.auth.resetPasswordForEmail`](/docs/reference/javascript/auth-resetpasswordforemail) to reset a user password. This method takes a parameter of `redirectTo`, which is an absolute URL to the update password page. This URL must be saved in your allowed [Redirect URLs](https://supabase.com/dashboard/project/_/auth/url-configuration) list found at [Authentication > Redirect Configuration](https://supabase.com/dashboard/project/_/auth/url-configuration).
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+```js
+await supabase.auth.resetPasswordForEmail('hello@example.com', {
+  redirectTo: 'http://example.com/account/update-password',
+})
+```
+
+</TabPanel>
+<TabPanel id="kotlin" label="Kotlin">
+
+Supabase provides a convenient method [`.sendRecoveryEmail`](/docs/reference/kotlin/auth-resetpasswordforemail) to reset a user password. This method takes a parameter of `redirectUrl` which we will use to pass an absolute URL to the update password page. This URL must be saved in your allowed [Redirect URLs](https://supabase.com/dashboard/project/_/auth/url-configuration) list found at [Authentication > Redirect Configuration](https://supabase.com/dashboard/project/_/auth/url-configuration) or it won't redirect the user.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:100]:
```suggestion
Supabase provides a convenient method [`.sendRecoveryEmail`](/docs/reference/kotlin/auth-resetpasswordforemail) to reset a user password. This method takes a parameter of `redirectUrl`, which is an absolute URL to the update password page. This URL must be saved in your allowed [Redirect URLs](https://supabase.com/dashboard/project/_/auth/url-configuration) list found at [Authentication > Redirect Configuration](https://supabase.com/dashboard/project/_/auth/url-configuration).
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+
+```kotlin
+supabase.gotrue.sendRecoveryEmail(
+    email = "hello@example.com",
+    redirectUrl = "http://example.com/account/update-password"
+)
+```
+
+If you are on one of the Kotlin targets that have built-in support for redirect url handling like Android, you may want to checkout [OAuth and OTP link verification](https://supabase.com/docs/reference/kotlin/initializing).
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:109]:
```suggestion
If you are on one of the Kotlin targets that have built-in support for redirect URL handling, such as Android, see [OAuth and OTP link verification](https://supabase.com/docs/reference/kotlin/initializing).
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+<Tabs
+  scrollable
+  size="small"
+  type="underlined"
+  defaultActiveId="js"
+  queryGroup="language"
+>
+<TabPanel id="js" label="JavaScript">
+
+To update the password we call the [`supabase.auth.updateUser`](/docs/reference/javascript/auth-updateuser) method and pass along the new password to this method.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:125]:
```suggestion
To update the password, call the [`supabase.auth.updateUser`](/docs/reference/javascript/auth-updateuser) method with the new password.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+
+```js
+await supabase.auth.updateUser({ password: new_password })
+```
+
+</TabPanel>
+<TabPanel id="kotlin" label="Kotlin">
+
+To update the password we call the [`.modifyUser`](/docs/reference/kotlin/auth-updateuser) method and pass along the new password to this method.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:134]:
```suggestion
To update the password, call the [`.modifyUser`](/docs/reference/kotlin/auth-updateuser) method with the new password.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 27]
```diff
@@ -0,0 +1,155 @@
...
+```
+
+</TabPanel>
+</Tabs>
+
+## Frequently asked questions
+
+### How are passwords stored?
+
+Supabase Auth uses [bcrypt](https://en.wikipedia.org/wiki/Bcrypt), a strong password hashing function to store a hash of user's passwords. You would not be able to impersonate a user by having access to this information. Each hash is accompanied by a randomly generated salt parameter that provides an additional layer of protection.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:149]:
```suggestion
Supabase Auth uses [bcrypt](https://en.wikipedia.org/wiki/Bcrypt), a strong password hashing function, to store hashes of users' passwords. Only hashed passwords are stored. You cannot impersonate a user with the password hash. Each hash is accompanied by a randomly generated salt parameter for extra security.
```

File: apps/docs/pages/guides/auth/passwords.mdx [hunk 28]
```diff
@@ -0,0 +1,155 @@
...
+
+This information is stored in the `encrypted_password` column of the `auth.users` table. The column's name is a misnomer (cryptographic hashing is not encryption), but is kept for historical and backward-compatibility reasons.
```
Comment (charislam) [inline apps/docs/pages/guides/auth/passwords.mdx:151]:
```suggestion
The hash is stored in the `encrypted_password` column of the `auth.users` table. The column's name is a misnomer (cryptographic hashing is not encryption), but is kept backward compatibility.
```

**Review by charislam — APPROVED** [review]
minor suggestions for conciseness and simplicity

File: apps/docs/pages/guides/auth/passwords.mdx (hunk 5 above)
Comment (hf) [reply apps/docs/pages/guides/auth/passwords.mdx:27]: Would actually prefer to bring up the leaked passwords first.

**Pull request merged**